UnitySDK/Assets/ML-Agents/Scripts/Sensor/WriteAdapter.cs

```diff
@@ -70,8 +78,30 @@ public void SetTarget(TensorProxy tensorProxy, int batchIndex, int channelOffset
         {
             set
             {
-                // Only TensorProxy supports 3D access
-                m_Proxy.data[m_Batch, h, w, ch + m_Offset] = value;
+                if (m_Data != null)
+                {
+                    var height = m_Shape[0];
+                    var width = m_Shape[1];
+                    var channels = m_Shape[2];
+
+                    Debug.AssertFormat(
+                        h >= 0 && h < height, "height value {0} must be in range [0, {1}]", h, height-1
+                    );
+                    Debug.AssertFormat(
+                        w >= 0 && w < width, "width value {0} must be in range [0, {1}]", w, width-1
+                    );
+                    Debug.AssertFormat(
+                        ch >= 0 && ch < channels, "channel value {0} must be in range [0, {1}]", ch, channels-1
+                    );
+
+                    // Math copied from TensorShape.Index(). Note that m_Batch should always be 0
+                    var index = m_Batch * height * width * channels + h * width * channels + w * channels + ch;
```

Comment:
Any reason we don't use `TensorShape`  instead of `int[]`?

Reply:
I would say it's a bit less user-friendly than `int[]` because you have to pass 1 to the batch size for our usage. It would also prevent us from using higher-dimensional observations (although I guess we wouldn't be able to run inference on that either). Plus I didn't know about it at the time :) 

I can take a stab at swapping in a separate PR - we could use TensorShape in the ISensor interface, but use int[] in the slightly-higher-level SensorBase interface.

https://github.cds.internal.unity3d.com/unity/UnityInferenceEngine/blob/master/UnityProject/Assets/Barracuda/Core/Tensor.cs#L35

Reply:
Makes sense.  My thoughts (which I didn't fully articulate) were that int[] would still be on the public interface of this, but could use TensorShape internally for things like computing the index.  Maybe it's a bit of overkill.

Reply:
Also, not a requirement for this to get merged.

Reply:
OK, let me try it in another PR.